### DISCUSSION TITLE
Wrap @Directive(...) into a helper decorator

### DISCUSSION THREAD

**Discussion by carlocorradini** [MichalLytek/type-graphql#1360] [Q&A] [answered]
It is possible to create a wrapper around `@Directive(...)` into a helper decorator to avoid useless repetition?

__Use Case__:
Imagine the following `@auth` directive:
```graphql
@auth(
  roles: [String!]! = [],
  permissions: [String!]! = [],
)
```

Using `@auth` with `@Directive(...)` is prone to errors and cause useless repetition:
```typescript
@Resolver(Example)
class ExampleResolver() {
  @Mutation(() => Example)
  @Directive(`@auth`)
  secureMutation() { // ... }

  @Query(() => Example)
  @Directive(`@auth(roles: ["ADMIN"], permissions: ["example:secureQuery"])`)
  secureQuery() { // ... }
}
```

It could be better if it's possible to create a wrapper around `@Directive(...)`:
```typescript
@Resolver(Example)
class ExampleResolver() {
  @Mutation(() => Example)
  @Auth
  secureMutation() { // ... }

  @Query(() => Example)
  @Auth({ permissions: ["example:secureQuery"] })
  secureQuery() { // ... }
}
```

Where the `Auth` decorator looks something like (simplified):
```typescript
type AuthArgs = {
  roles?: Roles[];
  permissions?: Permissions[];
};

function Auth(args?: AuthArgs) {
  const roles: Roles[] = args.roles ?? [Roles.ADMIN];
  const permissions: Permissions[] = args?.permissions ?? [];

  const rolesString = roles
     .map((role) => `"${role}"`)
     .join(',');
  const permissionsString = permissions
     .map((permission) => `"${permission}"`)
     .join(',');     

  return Directive(`@auth(roles: [${ rolesString }], permissions: [${ permissionsString }])`);
}
```

Is it possible? Thanks 🥳

**Comment by MichalLytek** [comment, marked as answer]
decorator is a function, you can see some examples of aliases for other decorators as well

**Reply by carlocorradini** [reply]
Like [this](https://stackoverflow.com/questions/65688664/alias-for-decorator-in-typescript) or?
Thanks

**Reply by carlocorradini** [reply]
@MichalLytek Sorry to bother you again, but I really struggle with this/

I've tried creating the following `@Auth(...)` decorator:
```ts
type AuthArgs = {
  roles?: Roles[];
  permissions?: Permissions[];
};

function Auth(args?: AuthArgs): PropertyDecorator & MethodDecorator & ClassDecorator;
function Auth(args?: AuthArgs): PropertyDecorator | MethodDecorator | ClassDecorator {
  return (targetOrPrototype, propertyKey, descriptor) => {
    const roles: Roles[] = args.roles ?? [Roles.ADMIN];
    const permissions: Permissions[] = args?.permissions ?? [];
    const authString = `
      @auth(
        roles: [${roles.map((role) => `"${role}"`).join(',')}],
        permissions: [${permissions.map((permission) => `"${permission}"`).join(',')}]
      )
    `;

    Directive(authString)(targetOrPrototype, propertyKey, descriptor);
  };
}
```

And applied like this:
```ts
@Resolver(Example)
class ExampleResolver() {
  @Query(() => Example)
  @Auth
  secureQuery() { // ... }
}
```

Unfortunately, the decorator is ignored and the operation succeeds even without an authorization token. If I simply use the plain `@Directive("@auth")` decorator, it works.

I really don't understand what I am doing wrong :/

Thanks 😥🥳

**Reply by carlocorradini** [reply]
Found the error. 
It's due to multiline template literal string. If you "inline" the string, it works! However, if you put a single newline, it does not work and no error message is shown (it's simply ignored). 
This behavior is known? and is due to `type-graphql` or `graphQL`?
Thanks
